component/web/security/src/main/java/org/exoplatform/web/security/hash/Argon2PasswordEncoder.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * This file is part of the Meeds project (https://meeds.io/).
+ * Copyright (C) 2023 Meeds Association
+ * contact@meeds.io
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+*/
+package org.exoplatform.web.security.hash;
+
+import org.apache.commons.codec.binary.Hex;
+import org.bouncycastle.crypto.generators.Argon2BytesGenerator;
+import org.bouncycastle.crypto.params.Argon2Parameters;
+import org.exoplatform.container.PortalContainer;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
+import org.exoplatform.web.security.codec.AbstractCodec;
+import org.exoplatform.web.security.codec.CodecInitializer;
+import org.exoplatform.web.security.security.TokenServiceInitializationException;
+import org.picketlink.idm.api.Attribute;
+import org.picketlink.idm.api.AttributesManager;
+import org.picketlink.idm.impl.credential.DatabaseReadingSaltEncoder;
+
+import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+
+public class Argon2PasswordEncoder extends DatabaseReadingSaltEncoder {
```

Comment:
```suggestion
public class Argon2PasswordEncoder extends HasingEncoder {
```

We dont want to extends DatabaseReadingSaltEncoder, as it use an algo from the JVM. 
We want a new encoder, so we should extends HashingEncoder only

Reply:
I didn't understand what you mean of using an algo from the JVM, i extended `DatabaseReadingSaltEncoder` which extends the `HashingEncoder` to not rewrite the `afterInitialize()` and the `getSecureRandomInstance(`) functions
and to still be able to set the secureRandomAlgoithm by server prop.
because if i extend directly the HashingEncoder then i will copy past the two functions from DatabaseReadingSaltEncoder so it doesn't make any sense

Reply:
It for class organization
The "old" encoder is DatabaseReadingEncoder.
We create a new one => so new class.
The common class between the two is HashingEncoder

As Argon2idEncoder have nothing common with DatabaseReadingEncoder except salt usage, I dont want that the new encoder extends the old one.